Replaced == with .equals() in String comparasion

siavash1986 opened the pull request:
replace == and != with .equals when comparing Java Strings. == tests for reference equality whereas .equals() tests for value equality. this code may lead to bugs, especially when "new String("   ")" used instead of String literals.

Here is the String .equals() method:
public boolean equals(Object anObject) {
        if (this == anObject) {
            return true;
        }
        if (anObject instanceof String) {
            String anotherString = (String) anObject;
            int n = count;
            if (n == anotherString.count) {
                int i = 0;
                while (n-- != 0) {
                    if (charAt(i) != anotherString.charAt(i))
                            return false;
                    i++;
                }
                return true;
            }
        }
        return false;
    }

siavash1986 committed:
This code seems to be using non-short-circuit which causes both side of the expression to be evaluated. the result can be inferred by only evaluating one side when using short-circuit operations.

siavash1986 committed:
Merge remote-tracking branch 'origin/master'

siavash1986 committed:
replace == with .equals when comparing Java Strings. == tests for reference equality whereas .equals() tests for value equality. this code may lead to bug, especially when "new String("   ")" used instead of String literals.

Here is the String .equals() method:
public boolean equals(Object anObject) {
        if (this == anObject) {
            return true;
        }
        if (anObject instanceof String) {
            String anotherString = (String) anObject;
            int n = count;
            if (n == anotherString.count) {
                int i = 0;
                while (n-- != 0) {
                    if (charAt(i) != anotherString.charAt(i))
                            return false;
                    i++;
                }
                return true;
            }
        }
        return false;
    }

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java:
```diff
@@ -366,7 +366,7 @@ private void upMyLocationOverlayLayers() {
 
     private void overlayMyLocationLayers() {
-        if (draggable & !readOnly) {
+        if (draggable && !readOnly) {
```
I’ve never seen anybody use bitwise `and` on a boolean, much less a capital-B Boolean. Surely that can’t be intentional. And why is draggable not a primitive? Further exploration may be warranted.

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/activities/GeoTraceGoogleMapActivity.java:
```diff
@@ -505,7 +505,7 @@ private void setupAutomaticMode() {
         Long timeDelay;
         TimeUnit timeUnitsValue;
-        if (units == getString(R.string.minutes)) {
+        if (units.equals(getString(R.string.minutes))) {
```
Since this could never have been true before, and now can be, have you exercised the true path?

siavash1986 replied:
== will still return true when Java pools Strings which seem like was the case in this scenario but String pooling is not a definitive behaviour and can not be guaranteed. I did exercise the true path and it worked as expected

dcbriccetti replied:
Ah, thanks for the reminder. I had forgotten that.

siavash1986 committed:
replaced Boolean objects to boolean primitives wherever object creation was not necessary.
Object creation is an expensive operation in regards to memory consumption, I debugged the application with primitive booleans and the behavior was expected. I recommend applying this principle to methods that require a Boolean object and possibly refactoring them into 2 methods since passing boolean to a method is considered a poor practice.

shobhitagarwal1612 commented:
@siavash1986 Since in this PR you have replaced `==` operator with `equals()`, so kindly include the PMD rule `UseEqualsToCompareStrings` currently excluded to avoid it in future as well. 
I did a fix of this in #1448, but since @dcbriccetti has already reviewed this one. So I am closing PR #1448 and you can include that PMD rule in this one by simply removing this [line](https://github.com/opendatakit/collect/blob/master/config/quality/pmd/pmd-ruleset.xml#L43)

@dcbriccetti @siavash1986 You both might also be interested in reviewing PR #1387

lognaturel commented:
Thanks, @shobhitagarwal1612 @dcbriccetti for the reviews!

@siavash1986 please limit each pull request to one change. This pull request is currently tracking your master branch which you are continuing to commit to. Please take a moment to read over the [contribution guide](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#submitting-a-pull-request), especially the section about using branches!

siavash1986 commented:
Thanks @lognaturel and @shobhitagarwal1612  for the heads up, I didn't realize this pull request is tracking my master branch

lognaturel commented:
I'll close this one and let you reopen two separate ones! You may be interested in the git cherry-pick or rebase commands so you can apply just the commits you want to a new branch.
